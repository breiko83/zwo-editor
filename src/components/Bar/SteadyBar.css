.segment {
  position: relative;
}
.bar {    
  border-radius: 10px;
  border: 1px solid white;
}

.comment {
  position: absolute;
  top: -60px;
  left: 105px;
  color: gray;
  padding: 5px;  
  font-size: 20px;
}