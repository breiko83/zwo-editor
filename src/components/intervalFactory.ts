import { v4 as uuidv4 } from 'uuid'
import { PaceType } from './Editor/PaceSelector';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "./Interval";

const defaultDuration = 300;
const defaultPower = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

export default {
  steady(interval: Partial<SteadyInterval>): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      duration: defaultDuration,
      power: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      duration: defaultDuration,
      startPower: defaultPower / 2,
      endPower: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      duration: defaultDuration,
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      cadence: defaultCadence,
      restingCadence: defaultCadence,
      repeat: 3,
      onDuration: 30,
      offDuration: 120,
      onPower: defaultPower,
      offPower: defaultPower / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
