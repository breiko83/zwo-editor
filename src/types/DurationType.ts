export type DurationType = "time" | "distance";