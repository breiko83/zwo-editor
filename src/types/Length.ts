export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
