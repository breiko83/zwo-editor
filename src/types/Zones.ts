export enum ZoneColor {
  GRAY = '#807F80',
  BLUE = '#0E90D4',
  GREEN = '#00C46A',
  YELLOW = '#FFCB00',
  ORANGE = '#FF6430',
  RED = '#E90000',
  WHITE = '#FFFFFF',
}

export const Zones = {
  Z1: {
    min: 0.1,
    max: 0.605
  },
  Z2: {
    min: 0.605,
    max: 0.763
  },
  Z3: {
    min: 0.763,
    max: 0.901
  },
  Z4: {
    min: 0.901,
    max: 1.053
  },
  Z5: {
    min: 1.053,
    max: 1.194
  },
  Z6: {
    min: 1.194,
    max: 2.0
  }
}

export const ZonesArray = [
  [Zones.Z1.min,Zones.Z1.max],
  [Zones.Z2.min,Zones.Z2.max],
  [Zones.Z3.min,Zones.Z3.max],
  [Zones.Z4.min,Zones.Z4.max],
  [Zones.Z5.min,Zones.Z5.max],
  [Zones.Z6.min,Zones.Z6.max],
]

export const zoneColor = (intensity: number): ZoneColor => {
  if (intensity < Zones.Z1.max) {
    return ZoneColor.GRAY;
  } else if (intensity < Zones.Z2.max) {
    return ZoneColor.BLUE;
  } else if (intensity < Zones.Z3.max) {
    return ZoneColor.GREEN;
  } else if (intensity < Zones.Z4.max) {
    return ZoneColor.YELLOW;
  } else if (intensity < Zones.Z5.max) {
    return ZoneColor.ORANGE;
  } else {
    return ZoneColor.RED;
  }
};
