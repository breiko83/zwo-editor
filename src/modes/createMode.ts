import { LengthType } from "../types/LengthType";
import { RunningTimes } from "../types/RunningTimes";
import { SportType } from "../types/SportType";
import BikeMode from "./BikeMode";
import RunMode from "./RunMode";
import { WorkoutMode } from "./WorkoutMode";

interface CreateModeParams {
  sportType: SportType;
  ftp: number;
  weight: number;
  runningTimes: RunningTimes;
  lengthType: LengthType;
}

// Creates appropriate type of mode for current sport type
export default function createMode({sportType, ftp, weight, runningTimes, lengthType}: CreateModeParams): WorkoutMode {
  switch (sportType) {
    case "bike": return new BikeMode(ftp, weight);
    case "run": return new RunMode(runningTimes, lengthType);
  }
}
