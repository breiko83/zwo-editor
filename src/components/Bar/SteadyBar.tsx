import React, { useState, useEffect } from 'react'
import './SteadyBar.css'
import { zoneColor, Zones } from '../../types/Zones'
import { Resizable } from 're-resizable'
import Label from '../Label/Label'
import { SteadyInterval } from '../../types/Interval'
import { intensityMultiplier } from './multipliers'
import { WorkoutMode } from '../../modes/WorkoutMode'

interface SteadyBarProps {
  interval: SteadyInterval;
  mode: WorkoutMode;
  onChange: (interval: SteadyInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
  showLabel: boolean;
}

const SteadyBar = ({interval, mode, ...props}: SteadyBarProps) => {
  const style = { backgroundColor: zoneColor(interval.intensity) }

  const [width, setWidth] = useState(mode.lengthToWidth(interval.length))

  const [height, setHeight] = useState(mode.intensityToHeight(interval.intensity))

  const [showLabel, setShowLabel] = useState(false)

  const [selected, setSelected] = useState(props.selected)

  useEffect(()=>{
    setSelected(props.selected) 
  },[props.selected])  

  const handleCadenceChange = (cadence: number) => {
    props.onChange({ ...interval, cadence })
  }

  const handleResizeStop = (dWidth: number, dHeight: number) => {
    setWidth(width + dWidth)
    setHeight(height + dHeight)

    notifyChange(dWidth, dHeight)
  }

  const notifyChange = (dWidth: number, dHeight: number) => {
    props.onChange({
      ...interval,
      length: mode.widthToLength(width + dWidth),
      intensity: mode.heightToIntensity(height + dHeight),
    })
  }

  return (
    <div className='segment'
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
      onClick={() => props.onClick(interval.id)}
      style={props.selected ? { zIndex: 10 } : {}}
    >
      {((selected || showLabel) && (props.showLabel)) &&
        <Label
          interval={interval}
          mode={mode}
          onCadenceChange={(cadence: number)=> handleCadenceChange(cadence)}
        />
      }
      <Resizable
        className='bar'
        size={{
          width: mode.lengthToWidth(interval.length),
          height: mode.intensityToHeight(interval.intensity),
        }}
        minWidth={3}
        minHeight={intensityMultiplier * Zones.Z1.min}
        maxHeight={intensityMultiplier * Zones.Z6.max}
        enable={{ top: true, right: true }}
        grid={[1, 1]}
        onResizeStop={(e, direction, ref, d) => handleResizeStop(d.width, d.height)}
        onResize={(e, direction, ref, d) => notifyChange(d.width, d.height)}
        style={style}
      >
      </Resizable>
    </div>

  );
}

export default SteadyBar