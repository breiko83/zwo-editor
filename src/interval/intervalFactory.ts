import { v4 as uuidv4 } from 'uuid'
import { PaceType } from '../types/PaceType';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "../types/Interval";
import { Distance, Duration, Length } from '../types/Length';
import { WorkoutMode } from '../modes/WorkoutMode';

const defaultDuration = new Duration(300);
const defaultDistance = new Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>, mode: WorkoutMode): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      onCadence: defaultCadence,
      offCadence: defaultCadence,
      repeat: 3,
      onLength: mode.lengthType === "time" ? new Duration(30) : new Distance(200),
      offLength: mode.lengthType === "time" ? new Duration(120) : new Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
