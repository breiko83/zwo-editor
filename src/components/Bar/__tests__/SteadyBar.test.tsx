import React from 'react';
import SteadyBar from '../SteadyBar';
import { Zones } from '../../../types/Zones'
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'
import intervalFactory from '../../../interval/intervalFactory';
import createMode from '../../../modes/createMode';
import { Duration } from '../../../types/Length';

const MockReact = React;

jest.mock('../../Label/Label', () => (props: any) =>
  MockReact.createElement("Label", props));

jest.mock('re-resizable', () => ({
  Resizable: (props: any) => MockReact.createElement("Resizable", props),
}));

jest.mock('uuid', () => ({
  v4: () => `mock-id`,
}));

describe('<SteadyBar>', () => {
  it('renders', () => {
    const mode = createMode({sportType: "bike", ftp: 250, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.steady({
      length: new Duration(50),
      intensity: Zones.Z3.min,
    }, mode);
  
    const component = renderer.create(
      <SteadyBar
        interval={interval}
        mode={mode}
        selected={false}
        showLabel={false}
        onChange={() => { }}
        onClick={() => { }}
      />
    );
  
    expect(component).toMatchSnapshot();
  });

  it('renders without Label when selected, but showLabel=false', () => {
    const mode = createMode({sportType: "bike", ftp: 250, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.steady({
      length: new Duration(50),
      intensity: Zones.Z3.min,
    }, mode);
  
    const component = renderer.create(
      <SteadyBar
        interval={interval}
        mode={mode}
        selected={true}
        showLabel={false}
        onChange={() => { }}
        onClick={() => { }}
      />
    );
  
    expect(component).toMatchSnapshot();
  });

  it('renders with Label when selected and showLabel=true', () => {
    const mode = createMode({sportType: "bike", ftp: 250, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.steady({
      length: new Duration(50),
      intensity: Zones.Z3.min,
    }, mode);
  
    const component = renderer.create(
      <SteadyBar
        interval={interval}
        mode={mode}
        selected={true}
        showLabel={true}
        onChange={() => { }}
        onClick={() => { }}
      />
    );
  
    expect(component).toMatchSnapshot();
  });
});
