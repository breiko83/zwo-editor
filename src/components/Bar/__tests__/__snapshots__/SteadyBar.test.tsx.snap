// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SteadyBar> renders 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <Resizable
    className="bar"
    enable={
      Object {
        "right": true,
        "top": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={500}
    minHeight={25}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 190.75,
        "width": 16.666666666666668,
      }
    }
    style={
      Object {
        "backgroundColor": "#00C46A",
      }
    }
  />
</div>
`;

exports[`<SteadyBar> renders with Label when selected and showLabel=true 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "zIndex": 10,
    }
  }
>
  <Label
    interval={
      Object {
        "cadence": 0,
        "id": "mock-id",
        "intensity": 0.763,
        "length": Duration {
          "seconds": 50,
        },
        "pace": 0,
        "type": "steady",
      }
    }
    mode={
      BikeMode {
        "ftp": 250,
        "lengthType": "time",
        "sportType": "bike",
        "weight": 75,
      }
    }
    onCadenceChange={[Function]}
  />
  <Resizable
    className="bar"
    enable={
      Object {
        "right": true,
        "top": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={500}
    minHeight={25}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 190.75,
        "width": 16.666666666666668,
      }
    }
    style={
      Object {
        "backgroundColor": "#00C46A",
      }
    }
  />
</div>
`;

exports[`<SteadyBar> renders without Label when selected, but showLabel=false 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "zIndex": 10,
    }
  }
>
  <Resizable
    className="bar"
    enable={
      Object {
        "right": true,
        "top": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={500}
    minHeight={25}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 190.75,
        "width": 16.666666666666668,
      }
    }
    style={
      Object {
        "backgroundColor": "#00C46A",
      }
    }
  />
</div>
`;
